src/hotspot/share/runtime/synchronizer.cpp

```diff
@@ -2336,185 +1389,55 @@ class HandshakeForDeflation : public HandshakeClosure {
-                                        &free_head_p, &free_tail_p,
-                                        &saved_mid_in_use_p);
-    } else {
-      local_deflated_count =
-          deflate_monitor_list_using_JT(&target->om_in_use_list,
-                                        &target->om_in_use_count, &free_head_p,
-                                        &free_tail_p, &saved_mid_in_use_p);
-    }
-    deflated_count += local_deflated_count;
-
-    if (free_head_p != NULL) {
-      // Move the deflated ObjectMonitors to the global free list.
-      guarantee(free_tail_p != NULL && local_deflated_count > 0, "free_tail_p=" INTPTR_FORMAT ", local_deflated_count=%d", p2i(free_tail_p), local_deflated_count);
-      // Note: The target thread can be doing an om_alloc() that
-      // is trying to prepend an ObjectMonitor on its in-use list
-      // at the same time that we have deflated the current in-use
-      // list head and put it on the local free list. prepend_to_common()
-      // will detect the race and retry which avoids list corruption,
-      // but the next field in free_tail_p can flicker to marked
-      // and then unmarked while prepend_to_common() is sorting it
-      // all out.
-#ifdef ASSERT
-      ObjectMonitor* l_next_om = unmarked_next(free_tail_p);
-      assert(l_next_om == NULL, "must be NULL: _next_om=" INTPTR_FORMAT, p2i(l_next_om));
-#endif
-
-      prepend_list_to_global_wait_list(free_head_p, free_tail_p, local_deflated_count);
-
-      OM_PERFDATA_OP(Deflations, inc(local_deflated_count));
-    }
-
-    if (saved_mid_in_use_p != NULL) {
-      // deflate_monitor_list_using_JT() detected a safepoint starting.
-      timer.stop();
-      {
-        if (is_global) {
-          log_debug(monitorinflation)("pausing deflation of global idle monitors for a safepoint.");
-        } else {
-          log_debug(monitorinflation)("jt=" INTPTR_FORMAT ": pausing deflation of per-thread idle monitors for a safepoint.", p2i(target));
-        }
-        assert(SafepointMechanism::should_process(self), "sanity check");
-        ThreadBlockInVM blocker(self);
-      }
-      // Prepare for another loop after the safepoint.
-      free_head_p = NULL;
-      free_tail_p = NULL;
-      if (log_is_enabled(Info, monitorinflation)) {
-        timer.start();
+  // Deflate any global idle monitors.
+  size_t deflated = deflate_monitor_list();
+  if (deflated > 0) {
+    // There are monitors that have been deflated. Unlink them from the in use list.
+    ResourceMark rm;
+    GrowableArray<ObjectMonitor*> delete_list((int)deflated);
+    _in_use_list.unlink_deflated(&delete_list);
+    HandshakeForDeflation hfd_hc;
+    Handshake::execute(&hfd_hc);
+    // After the handshake, safely free the monitors that were deflated in this cycle.
+    for (ObjectMonitor* monitor: delete_list) {
+      delete monitor;
+      if (SafepointMechanism::should_process(self)) {
+        // Honor block request.
+        ThreadBlockInVM tbivm(self);
```

Comment:
While we are honoring this safepoint, the ObjectMonitors that
are on the `delete_list` will not be on the`_in_use_list` so any
audit that runs during the safepoint might get confused.